Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,695 @@
+        """
+        Instantiates a sampler that outputs sequential indices. 
+
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices.
+        
+        Returns  
+        -------  
+        Sampler  
+            An instance of the Sampler class that will generate indices sequentially. 
+        
+        Example
+        -------
+        >>> sampler = Sampler.sequential(10)
+        >>> print(sampler.get_samples(5))
+        >>> print(next(sampler))
+        0
+        [0 1 2 3 4]
+        >>> print(sampler.next())
+        1
+        """
+        def function(x):
+            return x % num_indices
+
+        sampler = Sampler(function=function, num_indices=num_indices, sampling_type='sequential',  prob_weights=[
+            1/num_indices]*num_indices)
+        return sampler
+
+    @staticmethod
+    def _staggered_function(num_indices, stride, iter_number):
+        """Function that takes in an iteration number and outputs an index number based on the staggered ordering.
+        
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices.
+
+        stride: int
+            The stride between returned indices. The stride should be less than the num_indices.
+        
+        iter_number: int
+            The current iteration number of the sampler.         
+        
+        Returns
+        -------
+        int
+            The index to be outputted by the sampler corresponding to the `iter_number`
+        
+        """
+        iter_number_mod = iter_number % num_indices
+        floor = num_indices // stride
+        mod = num_indices % stride
+
+        if iter_number_mod < (floor + 1)*mod:
+            row_number = iter_number_mod // (floor + 1)
+            column_number = (iter_number_mod % (floor + 1))
+        else:
+            row_number = mod + (iter_number_mod - (floor+1)*mod) // floor
+            column_number = (iter_number_mod - (floor+1)*mod) % floor
+
+        return row_number + stride*column_number
+
+    @staticmethod
+    def staggered(num_indices, stride):
+        """
+        Instantiates a sampler which outputs in a staggered order. 
+
+        Parameters
+        ----------
+        num_indices: int
+            The sampler will select from a range of indices 0 to num_indices.
+
+        stride: int
+            The stride between returned indices. The stride should be less than the num_indices.
+
+        Returns  
+        -------  
+        Sampler  
+            An instance of the Sampler class that will generate indices in a staggered pattern. 
+            
+
+        Example
+        -------    
+        >>> sampler = Sampler.staggered(21,4)
+        >>> print(next(sampler))
+        0
+        >>> print(sampler.next())
+        4
+        >>> print(sampler.get_samples(5))
+        [ 0  4  8 12 16]
```

Comment:
Can you show an example with a large enough stride to show it turning to 1 and 2?